### PR TITLE
[FSDP][1/N] Update `summon_full_params(with_grads)` `None` gradient

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #87308 [FSDP][2/N] Fix grad zero vs. `None` edge case
* **#87314 [FSDP][1/N] Update `summon_full_params(with_grads)` `None` gradient**

This PR changes `summon_full_params(with_grads=True)`'s behavior to be such that if all ranks have `flat_param.grad = None`, then the original parameters will correctly have `orig_param.grad = None`. This is achieved with a preliminary all-reduce. Note that if a particular original parameter's gradient is `None` on all of the containing ranks, but not all ranks' `flat_param.grad = None`, then that particular gradient is still going to be set to zeros. This can be handled if desired in follow-up work.

### CODE REVIEW

File: torch/distributed/fsdp/flat_param.py [hunk 1]
```diff
@@ -845,25 +845,52 @@ def _free_low_precision_sharded_param(self):
 
     @torch.no_grad()
     def unshard_grad(self):
+        """
+        Unshards the handle's ``FlatParameter`` 's gradient. If all ranks have
```
Comment (zhaojuanmao) [inline torch/distributed/fsdp/flat_param.py:849]: nit: maybe add a comment that 'unshard_grad' is not used in critical path, only used in summon_full_params(), as it calls all_reduce and may have performance impact

**Review by zhaojuanmao — APPROVED** [review]
